STEP07 - Integration

JK-Kim4 opened the pull request:
## PR 설명
#### (기능구현과 리펙토링을 동시에 진행하다보니 커밋 하나에 커밋에 다른 파일이 끼어들어간것들이 자주있습니다!
 양해해주시면서 봐주시면 감사드리겠습니다!)
### Infrastructure Layer 작성
<!-- 해당 PR이 왜 발생했고, 어떤부분에 대한 작업인지 작성해주세요. -->
- 사용자(User) Infra 구현: e8387a1
- 포인트(Point) Infra 구현: cd3a993
- 상품(Item) Infra 구현: e6d06fc
-  쿠폰 (Coupon) Infra 구현: 108a36a effb1ff 9aea631
- 주문(Order) Infra 구현: f13c2ff
- 결제(Payment) Infra 구현: 1be3564
---
### 통합테스트 코드
- 사용자(User) 통합 테스트: 7813df9
- 포인트(Point) 통합 테스트: 605027a
- 상품(Item) 통합 테스트: 8e7287d
-  쿠폰 (Coupon) 통합 테스트: d1d02c8 3d1f735 04e9ebe 323f5e0
- 주문(Order) 통합 테스트: fc96e01 36e40c8 bd408a7 1b0e6ee
- 결제(Payment) 통합 테스트: d9954e6
---
#### 기타
- 포인트 수정: 5295139 78bea97 d9a54279
- 상품 수정: 011fd64 831a7e97
- 쿠폰 수정: 03969fd 752703d 76dc712 af7c3fe 04e9ebe 41141e5b
- 사용자 수정: fe59473 7813df9 41141e5b d9a54279
- 주문 수정: 6363985
- 상품 테스트 수정: 7458a8e f31b298
- 사용자 테스트 수정: df27337
- 주문 테스트 수정: fc96e01 e942051 36e40c8 bd408a7
- 포인트 테스트 수정: 605027a
- 쿠폰 테스트 수정: 3d1f735
---
## 리뷰 포인트
<!-- 
    리뷰어가 함께 고민해주었으면 하는 내용을 간략하게 기재해주세요.
    커밋 링크가 포함되면, 더욱이 효과적일 거예요! 
-->
### 테스트 코드 메서드 네이밍
```
테스트 코드 함수명을 지을 때 테스트 코드가 기능 명세를 나타낼 수 있는 네이밍을 고민하였습니다만 
뒤로 갈수록 기능구현이나 테스트 내용 작성에 신경쓰다보니 이상한 함수명들이 지어진것같습니다..
테스트 코드 네이밍이 어떻게 작성되어야 더 좋은 이름이 되는지 알고싶습니다!
```
###  통합 테스트 데이터 셋팅
```
통합 테스트를 작성하며 @BeforeEach로 테스트 데이터를 셋팅해주다보니 필요한 데이터가 상대적으로 많은 OrderFacadeTest.java의 경우는 
테스트 데이터 삽입 로직이 테스트 로직보다 훨씬 길게 작성되었는데요..더 효과적으로 초기 데이터를 셋팅하는 방법이 있는지 알고싶습니다!
(혹시 테스트 코드가 명확하게 작성되어있다면 @BeforeEach나 @AfterEach 등의 부분이 길어져도 상관이 없을까요?)
```
 [OrderFacadeTest.java](https://github.com/JK-Kim4/hhplus-be-ch2/blob/step07/src/test/java/kr/hhplus/be/server/integration/orderpayment/application/OrderFacadeTest.java)
## Definition of Done (DoD)
<!--
    DOD 란 해당 작업을 완료했다고 간주하기 위해 충족해야 하는 기준을 의미합니다.
    어떤 기능을 위해 어떤 요구사항을 만족하였으며, 어떤 테스트를 수행했는지 등을 명확하게 체크리스트로 기재해 주세요.
    리뷰어 입장에서, 모든 맥락을 파악하기 이전에 작업의 성숙도/완성도를 파악하는 데에 도움이 됩니다.
    만약 계획되거나 연관 작업이나 파생 작업이 존재하는데, 이후로 미뤄지는 경우 TODO -, 사유와 함께 적어주세요.

    ex:
    - [x] 상품 도메인 모델 구조 설계 완료 ( [정책 참고자료](관련 문서 링크) )
    - [x] 상품 재고 차감 로직 유닛/통합 테스트 완료
    - [ ] TODO - 상품 주문 로직 개발 ( 정책 미수립으로 인해 후속 작업에서 진행 )
-->

JK-Kim4 committed:
feat: 사용자 infra 구현

JK-Kim4 committed:
feat: 포인트 리펙토링

JK-Kim4 committed:
feat: 포인트 infra 구현

JK-Kim4 committed:
feat: 상품 리펙토링

JK-Kim4 committed:
feat: 상품 infra 구현

JK-Kim4 committed:
feat: 쿠폰 리펙토링

JK-Kim4 committed:
feat: 쿠폰 infra 구현

JK-Kim4 committed:
feat: 쿠폰 리펙토링

JK-Kim4 committed:
feat: 사용자 쿠폰 리펙토링

JK-Kim4 committed:
feat: 사용자 쿠폰 infra 구현

JK-Kim4 committed:
feat: 사용자 통합 테스트

JK-Kim4 committed:
feat: 상품 통합 테스트

JK-Kim4 committed:
feat: 상품 단위 테스트 수정

JK-Kim4 committed:
feat: 사용자 단위 테스트 수정

JK-Kim4 committed:
feat: 쿠폰 통합 테스트

JK-Kim4 committed:
feat: 쿠폰 단위 테스트 수정

JK-Kim4 committed:
feat: 사용자 쿠폰 통합 테스트 수정

JK-Kim4 committed:
feat: 사용자 쿠폰 단위 테스트 수정

JK-Kim4 committed:
feat: 주문 리펙토링

JK-Kim4 committed:
feat: 주문 infra 구현

JK-Kim4 committed:
feat: 주문 통합 테스트

JK-Kim4 committed:
feat: 주문 단위 테스트 수정

JK-Kim4 committed:
feat: 기타 수정

o3ozzvb commented on src/main/java/kr/hhplus/be/server/application/coupon/CouponFacade.java:
```diff
@@ -1,31 +1,57 @@
+        //쿠폰 조회
+        Coupon coupon = couponService.findById(criteria.getCouponId());
+        //사용자 조회
+        User user = userService.findById(criteria.getUserId());
+
+        //쿠폰 생성
+            //만료일 검증 / 잔여 수량 검증
+            //사용자 동일 쿠폰 소지 여부
+        UserCoupon userCoupon = coupon.issue(user);
+
+        //쿠폰 저장
+        userCouponService.save(userCoupon);
+        user.addUserCoupon(userCoupon);
+
         return UserCouponInfo.Issue.of(
-                issue.getId(),
-                issue.getCoupon().getId(),
-                issue.getUser().getId(),
-                issue.getIssueDateTime());
+                userCoupon.getId(),
+                userCoupon.getCoupon().getId(),
+                userCoupon.getUser().getId(),
+                userCoupon.getIssueDateTime());
     }
```
Coupon 도메인에서 User 정보를 받아서 발행하는 역할을 하도록 할 수도 있을 것 같아욥

o3ozzvb commented on src/main/java/kr/hhplus/be/server/domain/coupon/Coupon.java:
```diff
@@ -1,48 +1,67 @@
 
     public Coupon(CouponTemplate couponTemplate) {
-        this.id = couponTemplate.getId();
+        if(couponTemplate.getId() == null) {
+            this.id = couponTemplate.getId();
+        }
```
null이면 nul 대입..?

JK-Kim4 replied:
헐

o3ozzvb commented on src/main/java/kr/hhplus/be/server/domain/order/Order.java:
```diff
@@ -57,23 +94,18 @@ public void registerPayment(Payment payment) {
 
-    public void addOderItems(List<OrderItem> orderItems) {
-        this.orderItems = new OrderItems(orderItems, this);
+    public void addOrderItems(List<OrderItem> orderItems) {
+        this.orderItems = new OrderItems(orderItems);
     }
```
List<OrderItem>을 OrderItems 로 wrapping 하는 역할은 OrderItems 클래스에 있는 것이 좋지 않을까요?,, OrderItems 의 존재 의미가 사라지는 것 같아 보여요.

o3ozzvb commented on src/main/java/kr/hhplus/be/server/domain/order/OrderUser.java:
OrderUser 라는 클래스를 어떤 사유로 만드신 건지 궁금합니다,,

JK-Kim4 replied:
의도는 Payment 내부에서 Order나 User 메소드 실행이었는데, @Transient 애노테이션 사용방법을 잘 몰라서 올바르게 추가되지않은것같아요. Payment가 실제 Order와 User를 갖고있어야 잔액 차감이라든지, 재고차감 등 값 변환로직을 수행할수있을거같아서

저거 없애고 Order, User 직접 들고있는것으로 수정예정이에요!

JK-Kim4 committed:
feat: 사용자 리펙토링

- Point 소유 User 객체 이외 User로부터 포인트 사용 요청이 발생하지 않아 Point 조작 책임을 User 객체에게 부여

JK-Kim4 committed:
feat: 포인트 리펙토링

- Point 소유 User 객체 이외 User로부터 포인트 사용 요청이 발생하지 않아 Point 조작 책임을 User 객체에게 부여

JK-Kim4 committed:
feat: 사용자 테스트코드 리펙토링

- Point 소유 User 객체 이외 User로부터 포인트 사용 요청이 발생하지 않아 Point 조작 책임을 User 객체에게 부여

JK-Kim4 committed:
feat: 포인트 테스트코드 리펙토링

- Point 소유 User 객체 이외 User로부터 포인트 사용 요청이 발생하지 않아 Point 조작 책임을 User 객체에게 부여

JK-Kim4 committed:
feat: 쿠폰/사용자쿠폰 도메인 리펙토링

JK-Kim4 committed:
feat: 쿠폰/사용자쿠폰 infra 구현

JK-Kim4 committed:
feat: 쿠폰/사용자쿠폰 테스트코드

JK-Kim4 committed:
feat: 주문 리펙토링

JK-Kim4 committed:
feat: 주문 테스트 코드

JK-Kim4 committed:
feat: 기타 수정

JK-Kim4 committed:
feat: 주문 리펙토링

JK-Kim4 committed:
feat: 주문 테스트코드

JK-Kim4 committed:
feat: 결제 리펙토링

JK-Kim4 committed:
feat: 결제 infra 구현

JK-Kim4 committed:
feat: 결제 테스트코드

JK-Kim4 committed:
feat: 사용자 / 포인트 객체 생성 팩토리 메서드 추가

JK-Kim4 committed:
feat: 주문/결제 통합 테스트 코드

JK-Kim4 committed:
Merge branch 'step07' into feat/domain/refactoring

JK-Kim4 committed:
feat: 쿠폰 V2 문자열 제거

JK-Kim4 committed:
feat: 쿠폰 / 사용자 쿠폰 리펙토링

JK-Kim4 committed:
feat: 상품 리펙토링

JK-Kim4 committed:
feat: 상품 관련 테스트코드 리펙토링

JK-Kim4 committed:
Merge branch 'feat/domain/refactoring' into step07

# Conflicts:
#	src/main/java/kr/hhplus/be/server/domain/coupon/Coupon.java

JK-Kim4 committed:
feat: 사용자 / 주문 리펙토링

JK-Kim4 committed:
Merge branch 'feat/domain/refactoring' into step07

JK-Kim4 committed:
feat: 상품 테스트코드 수정

JK-Kim4 committed:
feat: 기타 수정

JK-Kim4 committed:
feat: OrderItem 버그 수정

- Item <-> OrderItem 간 연간관계가 @OneToOne으로 설정되어있어 동일 Item 주문 시 Duplication key 오류 발생
